kernel/work.c

```diff
@@ -0,0 +1,952 @@
+
+	/* Clear this first, so released high-priority threads don't
+	 * see it when doing things.
+	 */
+	atomic_clear_bit(&work->flags, K_WORK_CANCELING_BIT);
+
+	/* Move all pending cancels associated with this work item to
+	 * another list, under lock.  (We can't signal the release
+	 * while the lock is held because it's a reschedule point,
+	 * which would release the lock.)
+	 */
+	sys_slist_init(&cancels);
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&pending_cancels, wc, tmp, node) {
+		if (wc->work == work) {
+			sys_slist_remove(&pending_cancels, NULL, &wc->node);
+			k_sem_give(&wc->sem);
+		}
+	}
+
+	/* Release the lock.  No need to yield here, the work thread
+	 * does that.
+	 */
+	k_spin_unlock(&lock, key);
+}
+
+void k_work_init(struct k_work *work,
+		  k_work_handler_t handler)
+{
+	__ASSERT_NO_MSG(work != NULL);
+	__ASSERT_NO_MSG(handler != 0);
+
+	*work = (struct k_work)Z_WORK_INITIALIZER(handler);
+}
+
+/* Add a flusher work item to the queue.
+ *
+ * Invoked with work lock and queue lock both held.
+ *
+ * Caller must notify queue of pending work.
+ *
+ * @param queue queue on which a work item may appear.
+ * @param work the work item that is either queued or running on @p
+ * queue
+ * @param flusher an uninitialized/unused flusher object
+ */
+static void _queue_flusher(struct k_work_q *queue,
+			   struct k_work *work,
+			   struct z_work_flusher *flusher)
+{
+	bool in_list = false;
+	struct k_work *wn;
+
+	/* Determine whether the work item is still queued. */
+	SYS_SLIST_FOR_EACH_CONTAINER(&queue->pending, wn, node) {
+		if (wn == work) {
+			in_list = true;
+			break;
+		}
+	}
+
+	init_flusher(flusher);
+	if (in_list) {
+		sys_slist_insert(&queue->pending, &work->node,
+				 &flusher->work.node);
+	} else {
+		sys_slist_prepend(&queue->pending, &flusher->work.node);
+	}
+}
+
+/* Try to remove a work item from the given queue.
+ *
+ * Invoked with work lock held.
+ *
+ * Locks queue temporarily.
+ *
+ * @param queue the queue from which the work should be removed
+ * @param work work that may be on the queue
+ */
+static inline void queue_remove(struct k_work_q *queue,
+				struct k_work *work)
+{
+	if (atomic_test_and_clear_bit(&work->flags, K_WORK_QUEUED_BIT)) {
+		k_spinlock_key_t key = k_spin_lock(&queue->lock);
+		(void)sys_slist_find_and_remove(&queue->pending, &work->node);
+		k_spin_unlock(&queue->lock, key);
+	}
+}
+
+/* Potentially notify a queue that it needs to look for pending work.
+ *
+ * Invoked with queue lock held, possibly also work lock held.
+ *
+ * This may make the work queue thread ready, but as the queue lock is held it
+ * will not be a reschedule point.  Callers should yield after the lock is
+ * released where appropriate (generally if this returns true).
+ *
+ * @param queue to be notified.  If this is null no notification is required.
+ *
+ * @return true iff the queue was notified and woken, i.e. a reschedule is
+ * pending.
+ */
+static inline bool _notify_queue(struct k_work_q *queue)
```

Comment:
preceding function names with _ is not allowed in Zephyr, it violates some MISRA-C rule. these are static functions, can you come up with some other convention? I'm presuming you put the _ here to indicate locking semantics like is done in Linux

Reply:
That is a confusion that come from C standard. C99 standard says that identifiers that begin with underscore are reserved to C library / implementation, consequently mostly MISRA-C scanners consider it a violation.